test_loop_rw fails since 2.9 on s390x

The integration tests fail on Ubuntu/s390x since we updated to 2.9.0

https://objectstorage.prodstack4-5.canonical.com/v1/AUTH_77e2ada1e7a84929a74ba3b87153c0ac/autopkgtest-groovy/groovy/s390x/u/udisks2/20200610_155315_a918d@/log.gz

```
======================================================================
FAIL: test_loop_rw (__main__.Manager)
loop device R/W
----------------------------------------------------------------------
Traceback (most recent call last):
  File "src/tests/integration-test", line 616, in test_loop_rw
    self.assertEqual(block.get_property('id-label'), 'foo')
AssertionError: '' != 'foo'
+ foo
```

libblockdev was also updated from 2.23 to 2.24 but the udisks 2.8.4 tests are not failing with the updated library so it seems the issue is due to the udisks update itself

--- a/modules/bcache/udiskslinuxblockbcache.c
+++ b/modules/bcache/udiskslinuxblockbcache.c
@@ -263,6 +263,7 @@ udisks_linux_block_bcache_update (UDisksLinuxBlockBcache  *block,
   udisks_block_bcache_set_bypass_misses (iface, stats->bypass_misses);
 
 out:
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (iface));
   if (stats)
     bd_kbd_bcache_stats_free (stats);
   if (error)

--- a/modules/btrfs/udiskslinuxfilesystembtrfs.c
+++ b/modules/btrfs/udiskslinuxfilesystembtrfs.c
@@ -270,6 +270,7 @@ udisks_linux_filesystem_btrfs_update (UDisksLinuxFilesystemBTRFS *l_fs_btrfs,
   udisks_filesystem_btrfs_set_used (fs_btrfs, btrfs_info->used);
 
 out:
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (fs_btrfs));
   if (btrfs_info)
     bd_btrfs_filesystem_info_free (btrfs_info);
   if (error)

--- a/modules/iscsi/udiskslinuxiscsisessionobject.c
+++ b/modules/iscsi/udiskslinuxiscsisessionobject.c
@@ -355,6 +355,8 @@ udisks_linux_iscsi_session_object_update_iface (UDisksLinuxISCSISessionObject *s
   udisks_iscsi_session_set_lu_reset_timeout (iface, session_info.tmo.lu_reset_tmo);
   udisks_iscsi_session_set_recovery_timeout (iface, session_info.tmo.recovery_tmo);
   udisks_iscsi_session_set_tgt_reset_timeout (iface, session_info.tmo.tgt_reset_tmo);
+
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (iface));
 }
 
 static gboolean

--- a/modules/lsm/lsm_linux_drive.c
+++ b/modules/lsm/lsm_linux_drive.c
@@ -220,6 +220,8 @@ _fill_drive_lsm (UDisksLinuxDriveLSM  *drive_lsm,
   udisks_drive_lsm_set_min_io_size (std_drv_lsm, lsm_vol_data->min_io_size);
   udisks_drive_lsm_set_opt_io_size (std_drv_lsm, lsm_vol_data->opt_io_size);
   udisks_drive_lsm_set_raid_disk_count (std_drv_lsm, lsm_vol_data->raid_disk_count);
+
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (std_drv_lsm));
 }
 
 

--- a/modules/lvm2/udiskslinuxlogicalvolume.c
+++ b/modules/lvm2/udiskslinuxlogicalvolume.c
@@ -229,6 +229,8 @@ udisks_linux_logical_volume_update (UDisksLinuxLogicalVolume     *logical_volume
       logical_volume->needs_udev_hack = FALSE;
       g_free (dev_file);
     }
+
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (iface));
 }
 
 void
@@ -248,6 +250,7 @@ udisks_linux_logical_volume_update_etctabs (UDisksLinuxLogicalVolume     *logica
   udisks_logical_volume_set_child_configuration (iface,
                                                  udisks_linux_find_child_configuration (daemon,
                                                                                         uuid));
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (iface));
 }
 
 /* ---------------------------------------------------------------------------------------------------- */

--- a/modules/lvm2/udiskslinuxphysicalvolume.c
+++ b/modules/lvm2/udiskslinuxphysicalvolume.c
@@ -143,6 +143,8 @@ udisks_linux_physical_volume_update   (UDisksLinuxPhysicalVolume    *physical_vo
       udisks_physical_volume_set_size (iface, pv_info->pv_size);
       udisks_physical_volume_set_free_size (iface, pv_info->pv_free);
     }
+
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (iface));
 }
 
 /* ---------------------------------------------------------------------------------------------------- */

--- a/modules/lvm2/udiskslinuxvdovolume.c
+++ b/modules/lvm2/udiskslinuxvdovolume.c
@@ -177,6 +177,8 @@ udisks_linux_vdo_volume_update (UDisksLinuxVDOVolume         *vdo_volume,
 
   udisks_vdo_volume_set_compression (iface, vdo_info->compression);
   udisks_vdo_volume_set_deduplication (iface, vdo_info->deduplication);
+
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (iface));
 }
 
 /* ---------------------------------------------------------------------------------------------------- */

--- a/modules/lvm2/udiskslinuxvolumegroup.c
+++ b/modules/lvm2/udiskslinuxvolumegroup.c
@@ -136,11 +136,14 @@ udisks_linux_volume_group_update (UDisksLinuxVolumeGroup *volume_group,
                                   gboolean               *needs_polling_ret)
 {
   UDisksVolumeGroup *iface = UDISKS_VOLUME_GROUP (volume_group);
+
   udisks_volume_group_set_name (iface, vg_info->name);
   udisks_volume_group_set_uuid (iface, vg_info->uuid);
   udisks_volume_group_set_size (iface, vg_info->size);
   udisks_volume_group_set_free_size (iface, vg_info->free);
   udisks_volume_group_set_extent_size (iface, vg_info->extent_size);
+
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (iface));
 }
 
 /* ---------------------------------------------------------------------------------------------------- */

--- a/modules/lvm2/udiskslinuxvolumegroupobject.c
+++ b/modules/lvm2/udiskslinuxvolumegroupobject.c
@@ -442,6 +442,7 @@ block_object_update_lvm_iface (UDisksLinuxBlockObject *object,
 
   udisks_linux_block_lvm2_update (UDISKS_LINUX_BLOCK_LVM2 (iface_block_lvm2), object);
   udisks_block_lvm2_set_logical_volume (iface_block_lvm2, lv_obj_path);
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (iface_block_lvm2));
 }
 
 static void
@@ -456,6 +457,7 @@ lv_object_update_block_path (UDisksLinuxBlockObject       *block_object,
     {
       block_objpath = g_dbus_object_get_object_path (G_DBUS_OBJECT (block_object));
       udisks_logical_volume_set_block_device (UDISKS_LOGICAL_VOLUME (lv), block_objpath);
+      g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (lv));
     }
 }
 
@@ -695,6 +697,7 @@ update_vg (GObject      *source_obj,
   bd_lvm_vgdata_free (vg_info);
   lv_list_free (lvs);
 
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (object->iface_volume_group));
   g_object_unref (object);
 }
 

--- a/modules/vdo/udiskslinuxblockvdo.c
+++ b/modules/vdo/udiskslinuxblockvdo.c
@@ -251,6 +251,8 @@ do_refresh (UDisksBlockVDO *block_vdo,
 
   bd_vdo_info_free (bd_info);
 
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (block_vdo));
+
   return TRUE;
 }
 

--- a/modules/zram/udiskslinuxblockzram.c
+++ b/modules/zram/udiskslinuxblockzram.c
@@ -277,6 +277,7 @@ udisks_linux_block_zram_update (UDisksLinuxBlockZRAM    *zramblock,
 
   udisks_block_zram_set_active (iface, bd_swap_swapstatus (dev_file, &error));
 out:
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (iface));
   if (zram_info)
     bd_kbd_zram_stats_free (zram_info);
   if (error)

--- a/src/tests/integration-test
+++ b/src/tests/integration-test
@@ -116,10 +116,14 @@ class UDisksTestCase(unittest.TestCase):
     cd_device = ""
     client = None
     manager = None
+    system = False
 
     @classmethod
-    def init(cls, logfile=None):
+    def init(cls, logfile=None, system=False):
         """start daemon and set up test environment"""
+
+        # test against running system instance
+        cls.system = system
 
         if os.geteuid() != 0:
             print('this test suite needs to run as root', file=sys.stderr)
@@ -127,7 +131,7 @@ class UDisksTestCase(unittest.TestCase):
 
         # run from local build tree if we are in one, otherwise use system instance
         cls.daemon_path = os.path.join(srcdir, 'src', '.libs', 'udisksd')
-        if (os.access(cls.daemon_path, os.X_OK)):
+        if not system and (os.access(cls.daemon_path, os.X_OK)):
             print('Testing binaries from local build tree')
             cls.check_build_tree_config()
             # copy required system files
@@ -151,10 +155,12 @@ class UDisksTestCase(unittest.TestCase):
 
         (cls.device, cls.cd_device) = cls.setup_vdev()
 
-        # start polkit and udisks on a private DBus
-        cls.dbus = Gio.TestDBus()
-        cls.dbus.up()
-        os.environ['DBUS_SYSTEM_BUS_ADDRESS'] = cls.dbus.get_bus_address()
+        if not system:
+            # start polkit and udisks on a private DBus
+            cls.dbus = Gio.TestDBus()
+            cls.dbus.up()
+            os.environ['DBUS_SYSTEM_BUS_ADDRESS'] = cls.dbus.get_bus_address()
+
         # do not try to communicate with the current desktop session; this will
         # confuse it, as it cannot see this D-BUS instance
         try:
@@ -167,7 +173,7 @@ class UDisksTestCase(unittest.TestCase):
             cls.daemon_log = tempfile.TemporaryFile()
         atexit.register(cls.cleanup)
 
-        cls.start_daemon()
+        cls.start_daemon(system)
 
     @classmethod
     def cleanup(cls):
@@ -182,16 +188,18 @@ class UDisksTestCase(unittest.TestCase):
         cls.teardown_vdev(cls.device)
         cls.device = None
 
-        del os.environ['DBUS_SYSTEM_BUS_ADDRESS']
-        cls.dbus.down()
+        os.environ.pop('DBUS_SYSTEM_BUS_ADDRESS', None)
+        if cls.dbus:
+            cls.dbus.down()
 
     @classmethod
-    def start_daemon(cls):
-        assert cls.daemon is None
-        cls.daemon = subprocess.Popen([cls.daemon_path, '--replace'],
-                                      stdout=cls.daemon_log,
-                                      stderr=subprocess.STDOUT)
-        assert cls.daemon.pid, 'daemon failed to start'
+    def start_daemon(cls, system):
+        if not system:
+            assert cls.daemon is None
+            cls.daemon = subprocess.Popen([cls.daemon_path, '--replace'],
+                                          stdout=cls.daemon_log,
+                                          stderr=subprocess.STDOUT)
+            assert cls.daemon.pid, 'daemon failed to start'
 
         # wait until the daemon has started up
         timeout = 10
@@ -203,16 +211,18 @@ class UDisksTestCase(unittest.TestCase):
             cls.manager = cls.client.get_manager()
             timeout -= 1
         assert cls.manager, 'daemon failed to start'
-        assert cls.daemon.pid, 'daemon failed to start'
+
+        if not system:
+            assert cls.daemon.pid, 'daemon failed to start'
 
         cls.sync()
 
     @classmethod
     def stop_daemon(cls):
-        assert cls.daemon
-        os.kill(cls.daemon.pid, signal.SIGTERM)
-        os.wait()
-        cls.daemon = None
+        if cls.daemon:
+            os.kill(cls.daemon.pid, signal.SIGTERM)
+            os.wait()
+            cls.daemon = None
 
     @classmethod
     def sync(cls):
@@ -818,6 +828,9 @@ class FS(UDisksTestCase):
     def test_existing_udisks_mount_point(self):
         """fs: reuses existing udisks mount point"""
 
+        if self.system:
+            self.skipTest("Skipping when testing a system instance")
+
         self.mkfs('ext4', 'udiskstest')
         fs = self.udisks_filesystem()
 
@@ -834,7 +847,7 @@ class FS(UDisksTestCase):
 
         # restart daemon, mount again; this should use the same mount point as
         # before
-        UDisksTestCase.start_daemon()
+        UDisksTestCase.start_daemon(self.system)
         fs = self.udisks_filesystem()
         new_mount_path = fs.call_mount_sync(no_options, None)
         self.retry_busy(fs.call_unmount_sync, no_options, None)
@@ -1673,6 +1686,8 @@ class Polkit(UDisksTestCase, test_polkitd.PolkitTestCase):
     """Check operation with polkit."""
 
     def setUp(self):
+        if self.system:
+            self.skipTest("Skipping Polkit tests when testing a system instance")
         if DISTRO_VER == ("fedora", "26"):
             self.skipTest("Skipping hanging PolicyKit tests on Fedora 26")
 
@@ -1926,13 +1941,16 @@ def restore_files(restore_list, tmpdir):
 
 if __name__ == '__main__':
     argparser = argparse.ArgumentParser(description='udisks2 integration test suite')
+    argparser.add_argument('-s', '--system', dest='system',
+                           help='run the test against the system installed instance',
+                           action='store_true')
     argparser.add_argument('-l', '--log-file', dest='logfile',
                            help='write daemon log to a file')
     argparser.add_argument('testname', nargs='*',
                            help='name of test class or method (e. g. "Drive", "FS.test_ext2")')
     cli_args = argparser.parse_args()
 
-    UDisksTestCase.init(logfile=cli_args.logfile)
+    UDisksTestCase.init(logfile=cli_args.logfile, system=cli_args.system)
     if cli_args.testname:
         tests = unittest.TestLoader().loadTestsFromNames(
             cli_args.testname, __import__('__main__'))

--- a/src/udiskslinuxblock.c
+++ b/src/udiskslinuxblock.c
@@ -893,6 +893,7 @@ update_configuration (UDisksLinuxBlock  *block,
       configuration = g_variant_new ("a(sa{sv})", NULL);
     }
   udisks_block_set_configuration (UDISKS_BLOCK (block), configuration);
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (block));
 }
 
 #ifdef HAVE_LIBMOUNT_UTAB
@@ -1280,6 +1281,7 @@ udisks_linux_block_update (UDisksLinuxBlock       *block,
   update_mdraid (block, device, drive, object_manager);
 
  out:
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (block));
   if (device != NULL)
     g_object_unref (device);
   if (drive != NULL)

--- a/src/udiskslinuxdrive.c
+++ b/src/udiskslinuxdrive.c
@@ -950,6 +950,7 @@ udisks_linux_drive_update (UDisksLinuxDrive       *drive,
   ret = update_configuration (drive, object);
 
  out:
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (drive));
   if (device != NULL)
     g_clear_object (&device);
 

--- a/src/udiskslinuxdriveata.c
+++ b/src/udiskslinuxdriveata.c
@@ -339,6 +339,8 @@ udisks_linux_drive_ata_update (UDisksLinuxDriveAta    *drive,
   update_security (drive, device);
 
  out:
+  /* ensure property changes are sent before the method return */
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (drive));
   if (device != NULL)
     g_object_unref (device);
 
@@ -680,6 +682,9 @@ udisks_linux_drive_ata_refresh_smart_sync (UDisksLinuxDriveAta  *drive,
   ret = TRUE;
   /* update stats again to account for the IO we just did to read the SMART info */
   update_io_stats (drive, device);
+
+  /* ensure property changes are sent before the method return */
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (drive));
 
  out:
   g_clear_object (&device);

--- a/src/udiskslinuxencrypted.c
+++ b/src/udiskslinuxencrypted.c
@@ -237,6 +237,8 @@ udisks_linux_encrypted_update (UDisksLinuxEncrypted   *encrypted,
     update_metadata_size (encrypted, object);
 
   udisks_linux_block_encrypted_unlock (block);
+
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (encrypted));
 }
 
 /* ---------------------------------------------------------------------------------------------------- */
@@ -629,6 +631,9 @@ handle_unlock (UDisksEncrypted        *encrypted,
                                         udisks_block_get_device_number (block),
                                         g_udev_device_get_sysfs_attr (cleartext_device->udev_device, "dm/uuid"),
                                         caller_uid);
+
+  /* ensure property changes are sent before the method return */
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (encrypted));
 
   udisks_encrypted_complete_unlock (encrypted,
                                     invocation,

--- a/src/udiskslinuxfilesystem.c
+++ b/src/udiskslinuxfilesystem.c
@@ -277,6 +277,8 @@ udisks_linux_filesystem_update (UDisksLinuxFilesystem  *filesystem,
   if (! skip_fs_size)
     udisks_filesystem_set_size (UDISKS_FILESYSTEM (filesystem), get_filesystem_size (object));
 
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (filesystem));
+
   g_object_unref (device);
 }
 
@@ -1865,6 +1867,7 @@ handle_resize (UDisksFilesystem      *filesystem,
                                                  UDISKS_DEFAULT_WAIT_TIMEOUT);
 
   udisks_filesystem_set_size (filesystem, get_filesystem_size (UDISKS_LINUX_BLOCK_OBJECT (object)));
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (filesystem));
   udisks_filesystem_complete_resize (filesystem, invocation);
   udisks_simple_job_complete (UDISKS_SIMPLE_JOB (job), TRUE, NULL);
 

--- a/src/udiskslinuxloop.c
+++ b/src/udiskslinuxloop.c
@@ -187,6 +187,7 @@ udisks_linux_loop_update (UDisksLinuxLoop        *loop,
     }
   udisks_loop_set_setup_by_uid (UDISKS_LOOP (loop), setup_by_uid);
 
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (loop));
   g_object_unref (device);
 }
 

--- a/src/udiskslinuxmdraid.c
+++ b/src/udiskslinuxmdraid.c
@@ -512,6 +512,7 @@ udisks_linux_mdraid_update (UDisksLinuxMDRaid       *mdraid,
                                                                                 uuid));
 
  out:
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (mdraid));
   if (raid_data)
       bd_md_examine_data_free (raid_data);
   g_free (sync_completed);

--- a/src/udiskslinuxpartition.c
+++ b/src/udiskslinuxpartition.c
@@ -312,6 +312,8 @@ udisks_linux_partition_update (UDisksLinuxPartition   *partition,
   udisks_partition_set_is_container (UDISKS_PARTITION (partition), is_container);
   udisks_partition_set_is_contained (UDISKS_PARTITION (partition), is_contained);
 
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (partition));
+
   g_free (name);
   g_clear_object (&device);
   g_clear_object (&disk_block_object);

--- a/src/udiskslinuxpartitiontable.c
+++ b/src/udiskslinuxpartitiontable.c
@@ -146,6 +146,7 @@ udisks_linux_partition_table_update (UDisksLinuxPartitionTable  *table,
 
   udisks_partition_table_set_partitions (UDISKS_PARTITION_TABLE (table),
                                          partition_object_paths);
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (table));
 
 
   g_free (partition_object_paths);

--- a/src/udiskslinuxswapspace.c
+++ b/src/udiskslinuxswapspace.c
@@ -127,6 +127,7 @@ udisks_linux_swapspace_update (UDisksLinuxSwapspace   *swapspace,
     active = TRUE;
   udisks_swapspace_set_active (UDISKS_SWAPSPACE (swapspace), active);
 
+  g_dbus_interface_skeleton_flush (G_DBUS_INTERFACE_SKELETON (swapspace));
   g_object_unref (device);
 }
 

